[2단계 - 웹 자동차 경주] 도이(유도영) 미션 제출합니다.

yoondgu opened the pull request:
안녕하세요 시카 ! 🙂 주말 잘 보내고 계신가요?

이번에는 [README](https://github.com/woowacourse/jwp-racingcar/blob/eb3accb545f0a544c393671c844eddde8a053d33/README.md)에 미션 명세, 코드리뷰 반영 목록을 작성하였습니다.
지난 1단계에 남겨주신 피드백 중 반영 또는 학습한 사항에 대하여 목록에는 체크 표시를, 코멘트에는 답변을 달았습니다.

그 중 아직 더 학습과 고민이 필요한 것도 있지만, 질문드리고 함께 이야기하며 진행하고 싶어서 먼저 코드 리뷰 요청 드립니다 !

가장 큰 변경 사항은 아래와 같습니다.
  1. 서비스 계층 도입 [605cce4](https://github.com/woowacourse/jwp-racingcar/pull/105/commits/605cce4d30fbaae9b0ab8394b9b2f5b5f2068f70)
  2. DB 설계 변경 [e3b6ff6](https://github.com/woowacourse/jwp-racingcar/pull/105/commits/e3b6ff6bc68b437e3d9557b60f71b4105f2b5e82), [175b640](https://github.com/woowacourse/jwp-racingcar/pull/105/commits/175b640f1e2ad26de2cb7fbefe4006e0c82e4bb8)
  3. 도메인 객체에 대한 Dto 적용 (Car -> CarDto, JudgedCarDto) [b70225b](https://github.com/woowacourse/jwp-racingcar/pull/105/commits/b70225ba82296fef22b4778c210d2c249913d047)
  4. 공통 비즈니스 로직을 도메인에 통합 [a8fcdd4](https://github.com/woowacourse/jwp-racingcar/pull/105/commits/a8fcdd416ac1df7cfbdfa258480e43c28fa02216)
  
  콘솔 어플리케이션은 서비스 계층 없이 컨트롤러에서 racingGame.play()를 바로 실행하도록 했습니다.
  별도 데이터 저장 로직이 없기 때문입니다.
  만약 웹 어플리케이션처럼 Repository에 저장하고 이력을 조회하는 기능이 추가된다면 그때 만들어도 된다고 생각했습니다.
  웹 어플리케이션의 서비스 계층은 똑같이 racingGame.play()를 실행하지만, Dao 작업까지 수행해줍니다.

그리고 진행하며 어려움을 겪고 있는 부분은
### 1. DB 액세스를 위한 DTO 사용 괜찮을까요?
DB 설계 변경에 따라 같은 Car에 대한 Dto여도 아래와 같이 두 가지를 사용했습니다.

- CarDto : Car 객체의 필드 값을 전송 (name, position)
    - UI <-> 컨트롤러(요청정보 매핑 시), 컨트롤러 <-> 서비스 간 데이터 전송에 사용됨
- JudgedCarDto : Car 객체의 필드 값과, 우승자 여부를 전송 (name, position, isWinner)
    - 서비스 또는 콘솔 컨트롤러 <-> Dao 간 데이터 전송에 사용됨
    
name, position, isWinner 필드를 가지는 하나의 Dto만을 사용할 수도 있겠지만
그렇게 되면 컨트롤러 <-> 서비스 / 서비스 <-> Dao 간 같은 Dto를 사용하므로 결합도를 낮추기 위한 Dto의 의미가 퇴색된다고 보았습니다.

그렇지만 뭔가 어색하게 느껴져서 고민이에요.
Entity가 아닌 Dto를 이렇게 DB 액세스에 사용하는 방법이 일반적이지 않은지,
복잡도가 높아지는 방식인지 고민이 되어서 시카의 의견을 들어보고 싶습니다..!!

### 2. DB 설계 정규화 기준
우승자 정보는 게임에 참여한 자동차 중 "누가 우승했는지"에 대한 정보이므로
게임 이력 테이블의 winners 컬럼을 없애고
자동차 테이블이 is_winner 컬럼을 가지도록 했습니다.

테이블 간 정보의 불일치를 방지할 수 있으므로 좋다고 생각하지만,
이렇게 되니 게임 이력 테이블은 결국 count, created_at 만 저장하고
이력을 조회할 때 최신순으로 정렬하기 위해서만 사용됩니다.

불필요하게 테이블을 분리한 것일까요? 무조건 정규화를 많이 하는 게 좋은 것도 아닌 것 같습니다.
그래서 DB 설계를 할 때 시카는 어떤 기준으로 진행하시는지 여쭙고 싶습니다.

### 3. 테이블 간 조인이 필요할 때 Dao는 ?
Dao 클래스는 테이블 기준으로 나누는 게 좋다고 생각했습니다.
각 테이블의 CRUD를 구현하면 책임을 파악하기 쉽기 때문입니다.
그런데 Dao에서 조회 결과를 얻기 위해서 다른 테이블과의 조인이 필요할 때는 어떻게 하는 게 좋을까요..?
두 테이블을 조인하여 사용하는 별도의 Dao를 만들어야 할까요?

------

질문이 너무 길어졌네요..! 시간 내어 읽어주셔서 감사합니다 😆
그리고 Spring을 학습하는 레벨에서 DB, 아키텍처 관련 질문을 더 많이 드린 것 같은데 괜찮을까요..? 😅
먼저 리뷰 요청 드리지만, 남은 기간 동안 피드백 받으면서 예외처리도 적용하고 다른 부분들도 더 질문드리면서 개선해보겠습니다!!
Spring의 여러 기능을 탐색해보는 시간으로 생각하고 있어요!

그리고 [e3b6ff6](https://github.com/woowacourse/jwp-racingcar/pull/105/commits/e3b6ff6bc68b437e3d9557b60f71b4105f2b5e82) 이 커밋 메시지의 "문제 상황" 내용은 제가 ResponsEntity에서의 오류를 착각하고 잘못 작성한 내용입니다..ㅎㅎ

yoondgu committed:
chore: 콘솔 애플리케이션 코드 추가

- 1단계에서 누락된 기존 콘솔 어플리케이션 코드 추가

yoondgu committed:
docs: 미션 명세서 2단계 업데이트

- 2단계 반영하여 요구사항 목록 작성
- API 명세 작성
- 1단계 피드백 중 학습 후 반영할 목록 작성

yoondgu committed:
rename: 콘솔 애플리케이션 클래스명 수정

yoondgu committed:
docs: 1단계 피드백 학습 업데이트

- 해당 피드백에 대한 반영 내용은 실수로 6265f03d 에 커밋함

yoondgu committed:
refactor: 책임 분리를 위한 서비스 계층 도입

yoondgu committed:
docs: 1단계 피드백 학습 업데이트

yoondgu committed:
refactor: id를 반환하는 insert 메서드명 수정

yoondgu committed:
rename: Dto 이름 Dao, 테이블과 일관성있게 수정

yoondgu committed:
refactor: DB 테이블 설계 변경에 따른 전면 리팩터링

- 우승자 목록을 플레이 결과 테이블에 모두 문자열로 저장하는 대신, 자동차 별 우승 여부를 저장함.
- 우승 상태는 각 자동차가 가지는 정보이므로 테이블 간 불일치를 줄이기 위함.
- 우승자 테이블을 따로 만들지 않은 이유는 자동차 id 추가, 조인 등 불필요한 DB 작업을 늘리지 않기 위함.
- 가변객체인 Car에 대해 계층 별 DTO 도입
- Controller - Service : CarResponseDto
- Service - Dao : CarDto (엔티티는 아니나, 테이블 매핑을 위한 DTO)
- PlayResultDto 네이밍 PlayResponseDto 로 다른 Dto와 통일
- @RequestBody 관련 문제상황:
getter는 정의하지 않더라도 해당 어노테이션 내부동작에서 자동 생성해주는 것으로 알고 있는데,
PlayResponseDto, CarResponseDto 둘 중 하나라도 getter 정의하지 않으면
An illegal reflective access operation 으로 인해 컨트롤러 테스트 실패함.

yoondgu committed:
feat: Dao 게임 별 모든 자동차 정보 조회 구현

yoondgu committed:
feat: 게임 플레이 이력 조회 API 구현

yoondgu committed:
refactor: 자동차 경주 로직 도메인으로 이동

yoondgu committed:
refactor: 공통 비즈니스 로직을 최종 도메인 객체에 통합

- 이를 위해 기존 콘솔 게임의 도메인 코드 리팩터링 함께 진행함.
- 레이싱게임의 불필요한 초기화로직 삭제
- 레이싱게임 바로 플레이, 자동차 결과 반환받도록 공통 비즈니스 로직 통합함.
- 도메인에서 우승자 별도 조회 로직을 삭제하고 자동차 DTO 정보에 통합함. (데이터 일관성을 위함)
- Dto 변환하는 정적 팩토리 메서드명 Builder 대신 Converter로 수정함.

yoondgu committed:
refactor: 레이싱게임 빈 등록, 의존성 주입

yoondgu committed:
refactor: 웹 컨트롤러 파싱 구분문자 상수화

yoondgu committed:
refactor: final 키워드 누락 추가

yoondgu committed:
style: 검토할 부분 TODO 주석 추가

yoondgu committed:
test: 웹 어플리케이션 관련 테스트 DisplayName 작성

yoondgu committed:
rename: DB 테이블명 전면 수정

- CAR -> CARS
- PLAY_RESULT -> PLAY_RECORDS

yoondgu committed:
rename: Dto 클래스명 전면 수정, 불필요한 클래스분리 삭제

- CarResponseDto -> CarDto : Car 객체의 필드 값을 전송 (이름, 위치)
- CarDto 는 UI <-> 컨트롤러, 컨트롤러 <-> 서비스 간 데이터 전송에 사용됨
- CarDto -> JudgedCarDto : Car 객체의 필드 값과, 우승자 여부를 전송 (이름, 위치, 우승자 여부)
- JudgedCarDto 는 서비스 또는 콘솔 컨트롤러 <-> Dao 간 데이터 전송에 사용됨
- CarDtoConverter 삭제하고 CarDto 에 정적 메서드 추가

yoondgu committed:
refactor: 트랜잭션 어노테이션 위치 클래스로 변경, readOnly 속성 추가

yoondgu committed:
refactor: 이력 조회 관련 메서드명 수정, final 누락 추가

yoondgu committed:
refactor: 불필요한 빈 등록, 인스턴스 변수 삭제

yoondgu committed:
refactor: 일관성을 위한 메서드명 수정

dks301 commented on README.md:
```diff
@@ -38,3 +58,77 @@ Content-Type: application/json
+
+- [ ] 아래 피드백 학습 후 반영하기
+    - [x] [빈 public 생성자 선언하지 않고도 PlayRequestDto의 필드를 json mapping 할 수 있게 만드는 방법은?](https://github.com/woowacourse/jwp-racingcar/pull/82#discussion_r1165362349)
+        - [x] 기존 불필요한 디폴트 생성자 선언 코드 삭제
```
정리 👍
추가적으로 고민할 포인트들 1단계 코멘트에 달아드렸으니 보시고 지금 PR에 반영해주시면 될 것 같아요 ㅎㅎㅎ

dks301 commented on src/main/java/racingcar/dao/CarsDao.java:
```diff
@@ -0,0 +1,75 @@
+
+    // TODO 다른 테이블과 조인하는 쿼리를 해당 Dao에서 쓰는 게 맞을까?
+    public Map<Long, List<JudgedCarDto>> findAllCarsByPlayId() {
+        return jdbcTemplate.query(
```
> 그런데 Dao에서 조회 결과를 얻기 위해서 다른 테이블과의 조인이 필요할 때는 어떻게 하는 게 좋을까요..?

이건 개인적으로 정답이 없다고 생각하는 부분인데요...! 각각의 성능 or 유지보수 or 확장성 or etc 관점에서 장단점에 대해서 고민해보고 현재 상황에 맞게 선택해보면 어떨까요??

1. 말씀하신것처럼 테이블 기준으로 간단한 CRUD만 적용, join해야 하는 부분은 코드상에서 비즈니스 로직으로 처리한다면?
2. 테이블 기준 CRUD, join을 위한 DAO  따로 만들기
3. 현재 코드처럼 CarDAO 안에서 처리

yoondgu replied:
구체적으로 적용해볼 수 있는 상황들을 제시해주셔서 고민해보는 데에 큰 도움이 되었어요!! 감사합니다😆
현재 상황의 경우, 두 테이블을 join해 새로운 객체를 도출해낸다기보다는
Cars 테이블로부터 Car entity를 가져오는 상황이므로 CarDao에서 처리하는 게 충분히 자연스럽다고 생각해 3번 방식을 유지하기로 했습니다.

1번 방식도 좋다고 생각하지만, 도메인 간의 관계와 일치하도록(Play와 Car 간의 의존 관계 vs FK 참조 관계)
테이블 간의 관계를 구성하고 사용할 수 있게 하고 싶었습니다.

말씀 주신 것처럼 여러 고려사항, 관점에 따라 다르게 볼 수 있겠지만
책임 분리의 관점에서 보자면,
1번 방식처럼 "join해야 하는 부분을 비즈니스 로직으로 처리하는 것"은 DB 액세스를 위한 로직을 비즈니스 로직에서 책임지게 되어 좋지 않다고 생각합니다.
반면 3번 방식을 사용하면, 어떤 값들을 가져와 조립해야 하는지 알 필요 없어 좋다고 생각합니다.

그리고 이 내용과는 별개이지만
리뷰 주신 해당 메서드에서 최신순 조회를 하느라 조인을 했던 것이라서,
이러한 의도를 드러내기 위해
메서드명을 `findAllCarsByPlayId()`에서 `findAllCarsOrderByPlayCreatedAtDesc()` 으로 변경했습니다.

dks301 commented on src/main/java/racingcar/dao/CarsDao.java:
```diff
@@ -0,0 +1,75 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public void insert(final long id, final List<JudgedCarDto> cars) {
```
> 1. DB 액세스를 위한 DTO 사용 괜찮을까요?

근본적인 고민이 필요해보이네요...!
1. 결합도를 낮추면 어떠한 장점이 있어서 DTO를 사용하는걸까요?? 이번 미션에서의 웹 환경 + 기존 터미널 I/O 환경을 동시에 서비스하는 관점에서 고민해볼까요?
2. DB엑세스에서 dto 사용이 어색한건 Spring에서 제너럴한 형태이지 않아서 일텐데요...! 일반적인 spring data 환경에서 entity를 사용하는 이유가 무엇일까요??

yoondgu replied:
1. 각 관심사에 따라 책임을 분리할 수 있습니다. 예를 들어 Dao에서 도메인 객체를 직접 생성하거나 사용하면, DB 액세스 외의 일에 관여하는 것이 되기 때문입니다.

클라이언트/컨트롤러로 넘기는 DTO의 경우, 웹 환경과 기존 I/O 환경 간 표현 방식이 서로 달라져도
서비스, Dao 에서는 코드 변경이 일어나지 않는다는 점에서 낮아진 결합도가 유지보수성에 큰 장점이 되는 것 같습니다.

하지만 같은 관점에서, DB 액세스를 위한 DTO의 장점을 생각하면 앞서 말한 책임 분리 외에는 잘 모르겠습니다.
표현 계층을 위한 DTO와 달리, 영속 계층은  도메인이 바뀌면(예: 필드 - 컬럼 추가) 같이 바뀌는 게 자연스럽기 때문인 것 같아요.

이어 답해보자면
2. 
[DB 설계 정규화 기준](https://github.com/woowacourse/jwp-racingcar/pull/105#discussion_r1167795949)에 달아주신
코멘트를 바탕으로, DB 또한 도메인 객체를 기준으로 만들어지는 것이 객체지향적 설계를 위해서는 더 이상적이라고 생각하게 되었습니다.
그래서 앞서 말한 것처럼 영속계층이 도메인 변경에 따라 함께 변하는 것은 자연스럽기 때문에 entity를 사용하는 것 아닐까? 생각이 드네요..ㅎㅎ
그럼에도 도메인 entity가 여러 로직을 가지고 있다면, Dao에서 이를 직접 생성하고 사용하기보다는 연결을 한 번 끊어주는 게 좋을 것 같아요!

그래서 일단은 
- Cars 테이블 컬럼과 도메인 Car entity 필드의 일치화 (playId, name, position)
- DB 액세스를 위한 Dto 삭제, Car entity 사용
이렇게 수정한 뒤
Dao에서 사용하는 Car entity를 별도 패키지에 만들어 계층 간 결합도를 다시 낮춰주었습니다.

dks301 replied:
> 표현 계층을 위한 DTO와 달리, 영속 계층은 도메인이 바뀌면(예: 필드 - 컬럼 추가) 같이 바뀌는 게 자연스럽기 때문인 것 같아요.

1. 잘 짜여진 도메인과 DB설계가 보장된다면 도이가 이야기 한 부분도 맞는 말이지만, 그렇지 않은 경우를 고민해보면 어떨까요? ㅎㅎㅎ
예를들어, 처음의 isWinner를 생각해볼까요? isSecondaryWinner라는 요구사항이 생기면 필드-칼럼 추가가 필요할텐데요...!
같이 바뀌는게 자연스러울까요? 🤔

2. 지금은 h2를 사용하지만, 일반적으로 DB도 외부에 있는데요...! DB와 웹 환경에 어떤 차이가 있는지 고민해보면 어떨까요?? ㅎㅎ

yoondgu replied:
오늘 점심부터 지금까지 계속 고민을 했는데요.
명쾌하게 답하지 못하고 이렇게 답변이 길어졌다는 것은 아직 많이 부족하다는 뜻인 것 같습니다... ㅜㅜ
부족한 탓에 길어진 글을 올려 시카의 소중한 시간을 사용하게 해드려도 되나 고민이 되지만
그래도 의견 공유 드리고 싶어 코멘트 달아보아요..!

### 1. 영속 계층은 도메인과 같이 바뀌는 게 자연스럽다?
제가 했던 말을 정정하고 싶어졌습니다. ㅎㅎ

**"테이블과 매핑되는 도메인 객체에 새로운 속성이 추가되거나 타입이 바뀌는 경우"에 테이블의 컬럼이나 구조도 함께 바뀔 수 있다**
`ex. Car에 Team이 추가됨, Car의 position이 Double로 변경됨`
는 생각을 전달하고 싶어서 했던 말인데, 많이 부정확했던 것 같아요.

시카가 주신 질문의 의도를 제가 제대로 파악했는지 확신이 서지 않지만 😢 마저 대답해볼게요..!!

"2번 우승한 사람을 찾아야 한다"는 요구사항이 추가된 경우에는 영속계층이 함께 변할 수도, 아닐 수도 있다고 생각해요.

아래 적어본 것만 해도, 도메인, 영속 계층이 서로 불일치하면서 요구사항을 만족시키는 경우는 분명히 많은 것 같습니다.
항상 절대적으로 서로 일치하고, 무결하도록 보장된 도메인-DB 설계만 있는 것은 아니라는 것을 염두에 두어야겠습니다.

#### Cars 테이블이 isWinner 컬럼을 가지고 있다면 
1. 기존 DB, 도메인 entity 변경 없이 같은 name에 대한 isWinner가 참인 행을 집계해 조회한다. 또는
2. 각 Car의 우승 횟수를 저장하는 CarWinningCounts 테이블을 별도 만든다.
- Car 객체에 isSecondaryWinner 라는 필드를 추가해 모든 Car에 조회 결과를 반영할 수도 있고, `도메인-테이블 불일치`
- Cars 테이블과 조인해 2번 우승한 Car들만 조회해 반환할 수도 있다. `도메인-테이블 컬럼은 일치하지만, 하나의 테이블로 하나의 도메인 객체를 표현하지 못한다.`
등등 .. 
#### Cars 테이블이 isWinner 컬럼을 가지고 있지 않다면
1. 모든 데이터를 가져와 비즈니스 로직에서 각 playRecordId 별로 우승자를 구하고, 그 중 중복을 집계해 조회한다.  `도메인-테이블 일치하지만 비효율적`
또는 
2. isWinner 컬럼을 추가하고 위 방법 중 하나를 택한다.
등등 ..

### 2. DB와 웹 환경의 차이 - Spring data 에서는 entity를 사용하는 이유

웹 클라이언트와 서버 간 통신은 Media-Type, body의 내용 또는 형식 등은 달라질 수 있지만
무조건 HTTP 프로토콜로 이루어집니다.(맞을까요..?)
따라서 이에 대한 표준은 자세히 알 필요 없이 WAS로 추상화된 서비스를 제공받습니다.
다양한 형식으로 응답을 보낼 수 있습니다. (필드 중 특정 값만 담은 Dto, 원하는 형식으로 바꾼 Dto..) 

반면 DB의 경우, 이미 만들어진 프로젝트에서 지금 사용하는 것과 다른 DB로 교체하고 싶을 수도 있습니다.
그래서 웹 환경과 달리, 비즈니스 계층 - 영속성 계층 사이를 잇는 인터페이스가 중요한 것 같습니다.

결론적으로
웹 환경에 대한 통신과 달리, DB 액세스를 위해서는 "개발자가 인터페이스를 구현하는 입장이 된다"는 차이가 있다고 생각합니다.
인터페이스는 정해진 시그니처의 메서드를 정의하므로, 이에 대한 변경 가능성을 최소화하기 위해서
인자값으로 각 컬럼에 해당하는 값을 직접 받거나 비즈니스 로직의 영향을 받은 객체(Dto)를 사용하기보다는
테이블과 매핑되는 Entity를 받도록 표준을 정해두는 것이 좋다고 생각합니다. (JPA가 그런 것 같은데 맞나요?)

```java
  // 각 컬럼을 인자로 직접 받으면, 테이블의 컬럼이 바뀔 때마다 인터페이스를 수정해야 함
  void insert(String name, int position, boolean isWinner); 
  
  // Entity와 비슷해보이지만, 테이블과 매핑되지 않는다는 점에서 변경 가능성이 높음
  void insert(JudgedCarDto car); // name, position, isWinner
  
  // 일부 필드에 null을 넣어 전달할 수도 있지만, 테이블과 매핑됨
  void insert(Car car); // car의 어떤 필드를 사용할지는 세부 구현 사항임
```

구체적으로 설명할 만큼 잘 알지 못하는 것 같네요..
Spring data에 대해 아직 거의 모르기도 하고, 좀 더 경험과 지식을 쌓아야 정확히 대답할 수 있을 것 같습니다..! ㅜㅜ
제가 이번 레벨 테코톡 주제로 "jdbc vs sqlMapper vs ORM"을 맡았는데 열심히 공부해야겠습니다 ㅎㅎ

yoondgu replied:
오늘 점심부터 지금까지 계속 고민을 했는데요.
명쾌하게 답하지 못하고 이렇게 답변이 길어졌다는 것은 아직 많이 부족하다는 뜻인 것 같습니다... ㅜㅜ
부족한 탓에 길어진 글을 올려 시카의 소중한 시간을 사용하게 해드려도 되나 고민이 되지만
그래도 의견 공유 드리고 싶어 코멘트 달아보아요..!

### 1. 영속 계층은 도메인과 같이 바뀌는 게 자연스럽다?
제가 했던 말을 정정하고 싶어졌습니다. ㅎㅎ

**"테이블과 매핑되는 도메인 객체에 새로운 속성이 추가되거나 타입이 바뀌는 경우"에 테이블의 컬럼이나 구조도 함께 바뀔 수 있다**
`ex. Car에 Team이 추가됨, Car의 position이 Double로 변경됨`
는 생각을 전달하고 싶어서 했던 말인데, 많이 부정확했던 것 같아요.

시카가 주신 질문의 의도를 제가 제대로 파악했는지 확신이 서지 않지만 😢 마저 대답해볼게요..!!

"2번 우승한 사람을 찾아야 한다"는 요구사항이 추가된 경우에는 영속계층이 함께 변할 수도, 아닐 수도 있다고 생각해요.

아래 적어본 것만 해도, 도메인, 영속 계층이 서로 불일치하면서 요구사항을 만족시키는 경우는 분명히 많은 것 같습니다.
항상 절대적으로 서로 일치하고, 무결하도록 보장된 도메인-DB 설계만 있는 것은 아니라는 것을 염두에 두어야겠습니다.

#### Cars 테이블이 isWinner 컬럼을 가지고 있다면 
1. 기존 DB, 도메인 entity 변경 없이 같은 name에 대한 isWinner가 참인 행을 집계해 조회한다. 또는
2. 각 Car의 우승 횟수를 저장하는 CarWinningCounts 테이블을 별도 만든다.
- Car 객체에 isSecondaryWinner 라는 필드를 추가해 모든 Car에 조회 결과를 반영할 수도 있고, `도메인-테이블 불일치`
- Cars 테이블과 조인해 2번 우승한 Car들만 조회해 반환할 수도 있다. `도메인-테이블 컬럼은 일치하지만, 하나의 테이블로 하나의 도메인 객체를 표현하지 못한다.`
등등 .. 
#### Cars 테이블이 isWinner 컬럼을 가지고 있지 않다면
1. 모든 데이터를 가져와 비즈니스 로직에서 각 playRecordId 별로 우승자를 구하고, 그 중 중복을 집계해 조회한다.  `도메인-테이블 일치하지만 비효율적`
또는 
2. isWinner 컬럼을 추가하고 위 방법 중 하나를 택한다.
등등 ..

### 2. DB와 웹 환경의 차이 - Spring data 에서는 entity를 사용하는 이유

웹 클라이언트와 서버 간 통신은 Media-Type, body의 내용 또는 형식 등은 달라질 수 있지만
무조건 HTTP 프로토콜로 이루어집니다.(맞을까요..?)
따라서 이에 대한 표준은 자세히 알 필요 없이 WAS로 추상화된 서비스를 제공받습니다.
다양한 형식으로 응답을 보낼 수 있습니다. (필드 중 특정 값만 담은 Dto, 원하는 형식으로 바꾼 Dto..) 

반면 DB의 경우, 이미 만들어진 프로젝트에서 지금 사용하는 것과 다른 DB로 교체하고 싶을 수도 있습니다.
그래서 웹 환경과 달리, 비즈니스 계층 - 영속성 계층 사이를 잇는 인터페이스가 중요한 것 같습니다.

결론적으로
웹 환경에 대한 통신과 달리, DB 액세스를 위해서는 "개발자가 인터페이스를 구현하는 입장이 된다"는 차이가 있다고 생각합니다.
인터페이스는 정해진 시그니처의 메서드를 정의하므로, 이에 대한 변경 가능성을 최소화하기 위해서
인자값으로 각 컬럼에 해당하는 값을 직접 받거나 비즈니스 로직의 영향을 받은 객체(Dto)를 사용하기보다는
테이블과 매핑되는 Entity를 받도록 표준을 정해두는 것이 좋다고 생각합니다.
(지금은 직접 이 표준을 정하느라 고민이 많았지만 ORM을 쓸 경우 그 표준에 따라 DB에 액세스하게 될 것 같습니다. JPA가 그런 것 같은데 맞나요?)

```java
  // 각 컬럼을 인자로 직접 받으면, 테이블의 컬럼이 바뀔 때마다 인터페이스를 수정해야 함
  void insert(String name, int position, boolean isWinner); 
  
  // Entity와 비슷해보이지만, 테이블과 매핑되지 않는다는 점에서 변경 가능성이 높음
  void insert(JudgedCarDto car); // name, position, isWinner
  
  // 일부 필드에 null을 넣어 전달할 수도 있지만, 테이블과 매핑됨
  void insert(Car car); // car의 어떤 필드를 사용할지는 세부 구현 사항임
```

구체적으로 설명할 만큼 잘 알지 못하는 것 같네요..
Spring data에 대해 아직 거의 모르기도 하고, 좀 더 경험과 지식을 쌓아야 정확히 대답할 수 있을 것 같습니다..! ㅜㅜ
TMI지만 제가 이번 레벨 테코톡 주제로 "jdbc vs sqlMapper vs ORM"을 맡았는데 열심히 공부해야겠습니다 ㅎㅎ

dks301 replied:
1. 1번 답에 대해서는 생각하신 대로 정답이 없기 때문에 지금 하는 고민처럼 항상 고민후에 상황에 맞는 선택을 하면 될 것 같습니다! 💯
2. 간략하게 이야기해보면(자세히는 저도 찾아봐야해요😅) 서버 입장에서는 웹이든 DB든 사실 외부인데요...!
말씀하신 웹 환경에 맞는 인터페이스 표준을 Spring에서 제공합니다(WAS도 물론 맞고, 디스패쳐 서블릿,.. 등등) 따라서 우리는 컨트롤러 부터만 우리가 코드를 작성하면 되는거겠죠?

DB 영속성 계층은 어떨까요?? DB 영속성 계층또한 기본적인 인터페이스를 제공하지만, 웹처럼 완벽하게 표준화된 것이 아닌 말씀하신 것처럼 어느정도 개발자가 구현해야하는 연관관계(sql, schema, ...)가  필요하게 됩니다. 
여기서 객체지향과 관계형 데이터 베이스 사이 패러다임의 불일치가 생기가 되는데요.
보일러플레이트가 많아지고, 유지보수의 어려움(ex. 칼럼 수정 후 쿼리는 컴파일 에러가 없기에 런타임에 쿼리 오류를 파악한다던지)이 생기게 되고 이를 해결하기 위해서 Spring data에서는 추상화를 통해 Repository라는 인터페이스를 사용합니다 ㅎㅎ

> TMI지만 제가 이번 레벨 테코톡 주제로 "jdbc vs sqlMapper vs ORM"을 맡았는데 열심히 공부해야겠습니다 ㅎㅎ

와우 어려운 주제를 발표하시네요 ㅎㅎㅎ 영상 올라오면 저도 한 번 봐야겠어요😄 지금까지 같이 고민해본 부분이 공부하시는데 조금이나마 도움이 되길 바래요 👏👏

dks301 commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -0,0 +1,47 @@
+
+@Service
+@Transactional(readOnly = true)
+public class RacingCarService {
```
지금은 Service 계층을 DB 계층과 도메인 계층을 이어주는 비즈니스 계층으로 사용중이신데요...👍
어떤 관점에서 이렇게 하셨는지 설명해주실 수 있나요?? ㅎㅎ

yoondgu replied:
앗 저는 제 코드에서 Service 계층이 표현 계층(Controller)와 DB 계층을 이어주고 있고,
도메인의 경우 계층의 역할을 한다기보다는 Service가 비즈니스 로직을 수행하기 위해 도메인 객체를 도출해 사용하는 상황이라고 생각했습니다.
이 클래스에서 Dao는 필드로 의존하고 있는 반면,
도메인 객체은 RacingGame은 특정 비즈니스 로직인 playGame 메서드 스코프 내에서 객체로 생성해 사용하기 때문입니다.

혹시 DB 계층과 도메인 계층 간의 연결이라고 생각하시는 이유를 설명해주실 수 있을까요?!

yoondgu replied:
또, 표현 계층과 DB 계층을 이어주는 계층으로 사용한 이유를 설명드리자면
Layered Architecture 를 참고하여 서로 다른 관심사들을 레이어로 나누어 추상화하고자 했습니다!

dks301 replied:
1. RacingGame이 도메인이기 때문에 도메인에서의 게임 진행사항을 DB에 저장하기 때문에 이어준다고 표현했는데요...! 도이가 이야기하신 것을 토대로 정리하면 "도메인과 DB를 이어주고 그 응답을 표현 계층에 전달한다." 정도 겠네요 ㅎㅎㅎ 

2. Layered Architecture를 참고해서 관용적인 Repository라는 이름을 사용하셨는데요 ㅎㅎ RacingCarRepository 내부에서 CarEntity가 아닌 PlayRecord도 insert하는게 적적한지 고민해보면 어떨까요?

yoondgu replied:
1. 아 제가 이해가 부족했던 것 같아요! 설명해주셔서 감사합니다!! 

2. 
원래는 RacingCarRepository가 하나의 플레이에 대한 모든 정보를 관리하는 Repository라고 생각했습니다.
아래 세 줄의 코드를 하나의 로직으로 추상화할 수 있다고 생각했어요.

```java
	playDao.insert(PlayRecordEntity.build(racingGame.getCount()));
	long savedId = playRecordsDao.getLastId();
	carsDao.insert(toDaoEntities(savedId, racingGame.getCars));
```

이에 따라 playDao에 대한 insert를 따로 빼는 것은 본래 의도와는 다른 것 같습니다.

그러나 이런 고민이 발생한 이유는
정확히 어떤 도메인 객체에 대한 Repository인지 정의할 수 없는 클래스라는 점이고, 문제적인 상황이라 생각합니다.

만약 그럼에도 이 Repository를 사용하고 싶다면,
아래처럼 해당 Repository와 통신하는 별도의 객체를 정의해서 의도를 드러내는 게 좋아보입니다.
```java
  // GameResultRepository로 클래스명 변경
  pulbic void save(GameResultDto gameResult) {
	  playDao.insert(new PlayRecordEntity(gameResult.getCount()));
	  long savedId = playRecordsDao.getLastId();
	  carsDao.insert(savedId, toDaoEntities(gameResult.getCars));
  }
```

그렇지만 DB 액세스 부분은 Entity를 사용하기로 했는데 이에 대해 일관성이 깨지고, 
이득이 크지 않은 추상화를 위해 복잡도를 높이는 방식이라 차라리 Repository 계층을 없애는 게 낫다고 판단해
다시 Service가 직접 Dao를 관리하도록 변경하였습니다.

---------
#### DB 액세스 → 도메인 로직 수행 → DB 액세스 의 흐름 으로 수정
그리고 이와는 별개로, 여러 DAO 로직을 하나로 묶어 추상화하려고 한 시도 자체에 문제가 있었습니다.
도메인로직을 다 수행한 뒤, DB 업데이트를 한번에 해주었기 때문에 여러 로직을 하나로 묶을 수 있던 것인데,
이 과정에서 Car Entity를 부적절하게 사용하고 있음을 발견했습니다. 

Car Entity는 테이블 컬럼과 일치하는 필드를 가지게 만들어놓고
아래와 같이 id는 인자값을 따로 받고, 그 외 정보는 CarEntity를 통해 받고 있기 때문입니다.

```java
// interface CarsDao
void insert(Long id, List<CarEntity> cars); // CarEntity의 id에는 값을 담지 않음
```

객체 중심으로 로직을 수행하도록 한다고 생각하면, 이런 흐름이 더 나을 것 같아 아래와 같이 수정했습니다.

1. PlayRecordDao에서 플레이 정보를 저장하고, 새 id를 얻는다.
2. 새 id를 가진 도메인 객체 List<Car>를 생성한다.
3. 도메인 로직을 수행한다.
4. CarsDao에서 id를 가진 도메인 객체 정보를 받아 저장한다.

```java
    public PlayResponseDto playGame(final int count, final List<String> carNames) {
        playRecordsDao.insert(new PlayRecordEntity(count));
        final Long playRecordId = playRecordsDao.getLastId();
        final RacingGame racingGame = RacingGame.of(new RacingCarNames(carNames).createCars(playRecordId));
        
        racingGame.play(count);
        
        final List<Car> cars = racingGame.racingCars();
        carsDao.insert(EntityConverter.toDaoEntities(cars));
        return PlayResponseDtoConverter.from(cars, racingGame.findWinningCarNames());
    }
```

dks301 replied:
많은 고민을 하고 결정을 내리셨네요...! 결론 내신 부분에 대해서는 개인적으로도 같은 생각이에요 👍
지금까지 고민한 부분에서 더 나아가서
Spring data의 Repository가 왜 JdbcTemplate의 insert를 그대로 사용하지 않고 `Entity save(Entity entity)`로 메서드 시그니쳐를 바꾸었는지, 어떤 차이가 있는지 고민해보면 좋을 것 같아요 ㅎㅎㅎ

dks301 commented on src/main/resources/data.sql:
```diff
@@ -1,16 +1,17 @@
+    play_id   BIGINT      NOT NULL,
+    name      VARCHAR(10) NOT NULL,
+    position  INT         NOT NULL,
+    is_winner BOOLEAN     NOT NULL,
```
> 2. DB 설계 정규화 기준

무조건 정규화를 많이하는게 좋은 것이 아니라는 점에는 어느정도 동의해요 ㅎㅎㅎ
개인적으로 DB를 설계할 때는 먼저 도메인 관점에서 엔티티를 만들고, 거기에 따라 테이블을 만드는데요...!
지금까지 배웠던 객체지향에서의 객체 분리와 크게 다르지 않다고 생각해요.

예를들어, 지금과 같은 경우 테이블과 domain entity가 다른 필드를 가지고 있는데요...!
Car.class가 boolean isWinner를 가지는게 맞는지 고민해보면 어떨까요??

yoondgu replied:
우승자 여부는 기본 데이터로부터 파생되는 부차적인 정보(참여한 자동차들의 위치를 모두 비교해서 도출한 정보)이므로
isWinner를 테이블의 컬럼으로 두고, 또 객체의 필드로 가지는 것은
DB 조회 시에는 간편하지만 객체지향적 관점에서는 아쉬운 방법이라는 생각이 들었습니다.
 
그래서 해당 컬럼, 필드를 없애고,
플레이 이력 조회 시 테이블에서 가지고 있는 Car를 모두 조회한 뒤 도메인 객체 RacingGame를 생성해
우승자인 CarNames를 물어보도록 수정하였습니다.

dks301 replied:
이번 미션의 목표나. 학습적인 측면에서는 도이가 결론 내신 부분대로 처리하는게 좋은 것 같아요 👍

더 나아가서, DB의 비정규화도 객체지향 이랑 비슷한 관점에서 볼 수 있지 않을까요?

> ex. 원시값을 포장하라  or 불변 객체 사용

위와 같이 객체지향에서는 성능 적인 측면보다는 보다 읽기 쉬운 or 유연하고 확장하기 쉬운 코드를 지향하게 되는데요.
isWinner 처럼 비정규화 칼럼을 추가하는 것도 성능적인 부분이나 기타 요구사항이 생기는 경우 "정규화의 이점을 포기하면서 선택해야 할 수도 있다" 라고 생각해요 ㅎㅎㅎ

어떻게 생각하시나요??

yoondgu replied:
저도 말씀해주신 내용 모두 동감합니다 !!

지금은 많은 사람이 사용하거나, 데이터의 양이 많은 서비스가 아니므로
성능 이슈를 고려하지 않아도 되기 때문에, 학습적인 측면에서 나쁘지 않은 선택이라고 생각합니다..!!

하지만 성능적인 부분이 중요하다면, 지금의 DB 설계는 적절하지 않다고 생각해요.
현재 리팩터링한 것처럼 우승자 관련 정보를 DB에 저장하지 않으면,
같은 정보임에도 조회할 때마다 매번 우승자를 계산하는 로직을 반복해서 실행해야 하므로 자원 낭비가 심할 것입니다.
이에 대해서 말씀하신 것처럼 비정규화 컬럼을 추가함으로써 해당 프로젝트에서 얻어야 할 이점을 취할 수 있다고 생각합니다!!

dks301 reviewed (changes requested):
안녕하세요 도이, 리뷰어 시카입니다. 😄

코멘트 드린 내용에 대해서 많이 고민한게 느껴지네요! 👏👏
몇 가지 코멘트와 질문답변을 남겨드렸으니 확인해주세요~

추가로 궁금한 점이 생기면 언제든지 DM 주세요 ㅎㅎ

yoondgu committed:
refactor: insert, id 조회 메서드 분리

yoondgu committed:
docs: Spring 제공 기능 적용 목록 작성

yoondgu committed:
docs: 2단계 피드백 체크리스트 작성

yoondgu committed:
refactor: 기존 구현 코드 RacingCars 의존성 주입 적용

- DB 엔티티와 도메인 객체 간 일관성을 위한 리팩터링 진행중
- DB 조회 정보로 RacingCars 객체 생성 가능하게 하기 위함

yoondgu committed:
refactor: 테이블과 도메인 일치화, Dao에서 직접 entity 사용하도록 변경

- Dto, Entity 사용 검토
- Cars 테이블 컬럼과 도메인 Car entity 필드의 일치화 (playId, name, position)
- DB 액세스를 위한 Dto 삭제, Car entity 사용
- 우승자 여부는 기본 데이터로 파생되는 부차적인 정보이므로
 DB에 저장하지 않고, 플레이 이력 조회 시 도메인 객체를 생성해 도출하도록 함

yoondgu committed:
refactor: 도메인 엔티티와 Dao 엔티티 분리

- 로직이 없는 Dao 엔티티를 분리하여 도메인 - Dao 간 결합 끊기

yoondgu committed:
refactor: 여러 개의 Dao에 대한 로직을 Repository 클래스로 추상화

yoondgu committed:
refactor: Dao 구현체가 아닌 인터페이스에 의존하도록 변경

- 콘솔 어플리케이션도 저장 기능 구현 예정
- 하나의 서비스가 여러 저장 방식을 수행 가능하게 하기 위한 추상
- 스프링에서 생성자에 인터페이스 타입으로 정의했는데 어떻게 구현체를 찾아서 매핑하는걸까?

yoondgu committed:
feat: 콘솔 어플리케이션 저장 기능 구현

- 저장 기능 구현에 따라 서비스 계층 적용하여 중복 코드 제거

yoondgu committed:
feat: 콘솔 어플리케이션 이력 조회 기능 구현

- 단순 출력하도록 구현함
- 여러번 플레이 가능 + 이력 조회 요청하도록 컨트롤러 구현 필요

yoondgu committed:
feat: 콘솔 어플리케이션 여러 번 플레이 가능, 이력 조회 구현

yoondgu committed:
refactor: join 사용하는 Dao 메서드명 구체화

yoondgu committed:
refactor: 웹 어플리케이션 컨트롤러 예외처리 적용

yoondgu committed:
refactor: Dto에 Validation 어노테이션 추가

yoondgu committed:
refactor: Lombok 라이브러리 추가, 생성자, getter 어노테이션 적용

yoondgu commented:
안녕하세요 시카!
구체적으로 피드백을 남겨주셔서 막연했던 고민을 더 깊게 해볼 수 있었던 것 같아요.
정말 감사합니다.

이번에도 README의 피드백 체크리스트, 리팩터링 목록 업데이트와 코멘트 답변을 함께 진행했습니다.

다시 생각해보니, 콘솔 어플리케이션에서도 저장/조회 기능을 구현하고 이에 대한 중복도 제거하는 것이 이번 미션 목적에 맞는 것 같아 해당 부분도 추가 구현하였습니다.
그리고 그동안 미진했던 Spring에서 제공하는 다른 기능들도 조금 사용해보았는데요.

1. 예외 처리 : ExceptionHandler 사용, 예외 발생 정보를 응답 객체에 담아 전달하기
        - **어떤 예외까지 핸들링할 것인가? 모든 Exception을 한번에 받는 게 맞을지?** 에 대해 고민중입니다.
2. Dto에 Validation 어노테이션 추가
        - 도메인 검증 로직 이전에 일찍 검증할 수 있는 장점이 있다고 생각하여 적용했습니다.
        - **예외 메시지는 도메인 검증 로직과 동일하게 작성해야 할까? 구분할 필요가 있을까?** 에 대해 고민중입니다.
        - **Validation 어노테이션이 던지는 예외**는
        `BindingResult`인터페이스를 구현하는 `MethodArgumentNotValidException` 예외인데
        exception message를 가져오는 방식이 보통 예외랑 조금 다르더라구요..
        이 부분은 일단 메서드 오버로딩 후 처리했는데, 더 좋은 방식이 있을지도 고민중입니다.
3. Lombok 라이브러리 추가, 생성자, getter 어노테이션 적용
        - 외부 라이브러리에 너무 많이 의존하는 것은 좋지 않다고 생각하지만, Lombok은 일반적으로 많이 사용하는 것 같아 적용해보았습니다.
        - 필드가 변경될 때마다 기계적으로 생성자, getter 메서드를 다시 만들어야 했는데, 훨씬 편리해질 것 같습니다!
        
위 말씀드린 고민 내용, 테스트 격리 관련 내용을 추가 개선 예정입니다.
그치만 이전 리뷰 요청에 비해 코드 변경사항이 꽤 많은 것 같아 한 번 피드백 받고 싶어 요청드립니다.. ㅎㅎ
이번에도 잘 부탁드립니다!! 감사합니다 😁

dks301 commented on src/main/java/racingcar/repository/dao/entity/CarEntity.java:
```diff
@@ -0,0 +1,50 @@
+
+public class CarEntity {
+
+    private final Long playId;
```
id를 통해서 간접참조 👍

playId vs playRecordsId 에 대해서 고민해보면 어떨까요??

yoondgu replied:
저도 그게 더 정확해서 좋다고 생각합니다! 테이블 컬럼명과 함께 변경하겠습니다.

dks301 commented on src/main/java/racingcar/repository/dao/JdbcPlayRecordsDao.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class JdbcPlayRecordsDao implements PlayRecordsDao {
```
PlayRecordDao가 있으니 entity를 추가해보면 어떨까요?? ㅎㅎ

yoondgu replied:
딱 하나의 값 (count)만 넣는데 Entity가 필요한가? 생각했다가
DB에서 entity를 쓰면 좋은 점이 이 [코멘트](https://github.com/woowacourse/jwp-racingcar/pull/105#discussion_r1173799403) 2번에서 정리가 되어 동의하게 됐습니다 ㅎㅎ
그리고 특정 값만 넣어 전달하고 싶은 경우를 고려해 builder 패턴을 적용해보았는데
여러 개의 생성자가 필요한 상황이 아니라 불필요하다고 판단해 다시 없애기로 했습니다.

dks301 reviewed (changes requested):
안녕하세요 도이, 리뷰어 시카입니다. 😄

코멘트 반영하면서 잘 구현해주셨네요 👍
추가 리팩토링에 고민해볼만한 내용 코멘트 남겨드렸으니 확인부탁드려요.

궁금한 점이 생기면 언제든지 DM 주세요 ㅎㅎ

---

추가로 고민중이신 부분은 정답이 없으니 충분히 고민해보시고 어떻게 결론내렸는지 알려주시고 같이 이야기 해보면 좋을 것 같아요! 😄

yoondgu committed:
style: 피드백 관련 고민거리 TODO 주석 추가

yoondgu committed:
refactor: 플레이 이력 테이블에 대한 엔티티 정의

- 여러 개의 생성자가 필요한 엔티티이므로 빌더 패턴 적용

yoondgu committed:
refactor: FK 컬럼명 참조하는 테이블에 일치하도록 변경, 관련 변수명 통일

yoondgu committed:
refactor: 예외 요청오류와 서버오류 구분하여 처리

yoondgu committed:
refactor: Repository 계층 삭제, 도메인-DB 연동 로직 변경

yoondgu committed:
refactor: 불필요한 빌더 패턴 제거

yoondgu committed:
build: repository 계층 삭제에 따른 패키지 구조 변경

yoondgu committed:
refactor: 메서드명 오타 수정

yoondgu committed:
refactor: 자동차 객체 생성 정적 팩토리 메서드 적용

yoondgu committed:
refactor: 컨트롤러 단의 예외 처리 로직 리팩터링

- Validation을 사용한 Dto 검증 예외: MethodArgumentNotValidException 의 에러메시지 추출
- handle 메서드에서 body에 반환할 정보에만 집중할 수 있도록 @ResponseStatus 사용

yoondgu committed:
refactor: RestControllerAdvice 사용하여 ExceptionHandler 전역 처리

- 예외 처리 페이지로 리다이렉트 시키려면 ControllerAdvice 를 사용할 것.
- 그러나 현재 요구사항이 아니라고 판단하여 예외 관련 정보만 반환함.

yoondgu committed:
refactor: 커스텀 예외 사용, 예측하지 못한 예외는 로그 처리

- 모든 Exception을 처리하면, 원하지 않는 정보를 사용자에게 노출할 위험이 있음.
- 커스텀 예외 이외의 예외에 대해서는 로그를 남기고 정해진 메시지만 응답으로 반환.

yoondgu committed:
refactor: Dao 테스트 JdbcTest 사용하도록 변경

yoondgu committed:
refactor: SpringBootTest 시 DB와 테스트 격리
- test 디렉토리에 테스트용 DB와 연결하도록 application.properties 작성

yoondgu commented:
안녕하세요 시카! 도이입니다 😀
추가 고민거리 던져주신 부분들에 대하여 답변을 달았습니다.
제가 아직 부족한 점이 많은 만큼 생각을 정리하는 데에도 오래 걸렸고, 의견을 전달드리는 데 있어서 명확하게 답하지 못하는 것 같아요.
그래서 답변들이 길기도 길고 읽기 불편하실 수도 있을 것 같아 그 점 양해 부탁드리고,
그럼에도 함께 대화 나눠주셔서 미리 감사드려요. 🥺

피드백을 받고 변경하기로 결정한 부분들을 다시 리팩터링하였고,
지난 리뷰 요청에서 고민 중이라고 했던 부분들도 반영해보았어요!
[README.md](https://github.com/yoondgu/jwp-racingcar/blob/step2/README.md)의 리팩터링 목록에 각 사항에 대한 의견을 함께 작성했으니 같이 봐주시면 감사하겠습니다.

yoondgu committed:
refactor: PlayRecords 테이블에 대한 Entity 누락 컬럼 추가

yoondgu committed:
refactor: 예외 응답객체 형식 통일

dks301 reviewed (approved):
안녕하세요 도이, 리뷰어 시카입니다. 😄

이번 미션에서 스프링에 대해 정말 많은 고민을 하고 공부하신 것 같아요 💯👏
특별히 개선할 부분이 없어 이번 미션은 이만 머지하겠습니다. 
수고 많으셨어요 ㅎㅎ

추가로 궁금한 점이 생기면 언제든지 DM 주세요 ㅎㅎ
